Build failure: rustc on LLVM stdenv

### Steps To Reproduce

Steps to reproduce the behavior:
1. build *rustc* with LLVM stdenv.
```nix
let
  nixpkgs = builtins.getFlake "github:NixOS/nixpkgs/nixos-unstable";
  pkgs = import nixpkgs {
    localSystem = "x86_64-linux";
    crossSystem = {
      system = "x86_64-linux";
      useLLVM = true;
      linker = "lld";
    };
  };
in pkgs.pkgsBuildHost.rustc
```

### Build log

[Here](https://gist.github.com/thenhnn/3da34df9941fd136169b997409c0d26c)

### Notify maintainers

@figsoda @Mic92 @tjni @winterqt @zowoq

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.30, NixOS, 24.05 (Uakari), 24.05.20240507.b211b39`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/f0ssy3p6898x9k880l23mhd0fl6rvrb2-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

The issue is that rust is unconditionally trying to link libgcc_s, but under pkgsLLVM there are no gcc builtins.

The relevant error is:
```
x86_64-unknown-linux-gnu-rustc>   = note: LC_ALL="C" PATH="/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/bin:/build/rustc-1.77.2-src/build/x86_64-unknown-linux-gnu/stage0/bin:/nix/store/md6hh4rrcrf99nssvcam3qaqs3skj086-gcc-wrapper-13.2.0/bin:/nix/store/9hgsinpfgyvsd92v0wlvmxv9wnaal68r-gcc-13.2.0/bin:/nix/store/nkdgysx90pln2l562l7lsk021hbxjdg8-glibc-2.39-31-bin/bin:/nix/store/ddypgcb25zgi0gwvklxlzazjrh9s63dg-coreutils-9.5/bin:/nix/store/m3gq2f6iwh3g67zrs27m7w82b37l8gf8-binutils-wrapper-2.41/bin:/nix/store/nd7ass1isid4vswpkfm82za4mcdwhzsm-binutils-2.41/bin:/nix/store/mil8a9jgadf9jjd4a9fw88l9kqxzdd49-pkg-config-wrapper-0.29.2/bin:/nix/store/l0yzhkqb9615ifq2gcl2r9i6544qisn1-file-5.45/bin:/nix/store/k3701zl6gmx3la7y4dnflcvf3xfy88kh-python3-3.11.9/bin:/nix/store/xwp10hiip3yyv879wqmxrbdhxal4khgh-rustc-wrapper-1.77.2/bin:/nix/store/pb69k0zwhxg6kf9ifz13x1wjs6a9vmf0-cmake-3.29.2/bin:/nix/store/zwc8msfmciilw22rn4rrc80n92ja20f2-which-2.21/bin:/nix/store/g5lrpqkk0ryyn784ra7lgyx8fnvh45sn-remove-references-to/bin:/nix/store/3iiml9pwff4y6zsis1gp4apbry1h1wfr-xz-5.4.6-bin/bin:/nix/store/r25jqybf7xs83k596glpj98mq5g8lx46-lndir-1.0.5/bin:/nix/store/8g7b1jkj5fpq1xcy5jgxfn9fxjwhxzvb-patchelf-0.15.0/bin:/nix/store/pmg30939m4ymrqy83fzyzrlhnzd69dl4-openssl-3.0.13-bin/bin:/nix/store/kyrawb1mqn0r10n6sa24f4j5hnl8kmrc-llvm-17.0.6-dev/bin:/nix/store/2nj2qa56cnq6shqskd9ixipc01s302wb-ncurses-6.4-dev/bin:/nix/store/jrrnf6kak9fbfm910506ppws50d33p9f-ncurses-6.4/bin:/nix/store/wyaplww8a5ir8j1d40p7873kzbgi3b4k-llvm-17.0.6/bin:/nix/store/ddypgcb25zgi0gwvklxlzazjrh9s63dg-coreutils-9.5/bin:/nix/store/y21vimynf3h666dvi4z8abzphw8ibibj-findutils-4.9.0/bin:/nix/store/4crhxg3dmy1n16h55rk6banv5xssmsyp-diffutils-3.10/bin:/nix/store/4cps736z7in3d37qc801lwv9z0ib67ps-gnused-4.9/bin:/nix/store/wy37jk2hirzqzx0666w1849kjdgzdam6-gnugrep-3.11/bin:/nix/store/a2ffqcqzbp8k4fmkyhpjq2km0i9wr8d8-gawk-5.2.2/bin:/nix/store/4m4vnjdy44s4sb8hlv4qsph492zz3rly-gnutar-1.35/bin:/nix/store/hragmx5p3izmwinv4441hj5706sxy4xs-gzip-1.13/bin:/nix/store/5giwsjx504zrgzl7cjzxm0s1423wndif-bzip2-1.0.8-bin/bin:/nix/store/qwcdasr63dfp74g7djazx48fx9s28i7a-gnumake-4.4.1/bin:/nix/store/h3bhzvz9ipglcybbcvkxvm4vg9lwvqg4-bash-5.2p26/bin:/nix/store/lj804cdk0fcf9yhgqrvgaqv2dc3lmfgp-patch-2.7.6/bin:/nix/store/3iiml9pwff4y6zsis1gp4apbry1h1wfr-xz-5.4.6-bin/bin:/nix/store/kaiq352476xq7zb1p06yranc5j3jjlz2-file-5.45/bin" VSLANG="1033" "/nix/store/z72l2dr5p8702z60iq27av6dm2fyl7y8-x86_64-unknown-linux-gnu-clang-wrapper-17.0.6/bin/x86_64-unknown-linux-gnu-clang" "-m64" "/build/rustcM4B7YH/symbols.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.00.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.01.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.02.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.03.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.04.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.05.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.06.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.07.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.08.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.build_script_build.20d8daabb4bb3e69-cgu.09.rcgu.o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb.4bmi8428vs7r72qu.rcgu.o" "-Wl,--as-needed" "-L" "/build/rustc-1.77.2-src/build/bootstrap/debug/deps" "-L" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-40bfb90c61fea8bb.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-95e6fcda3c282ef0.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-4c75151b6abbc292.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-35d5d497b8d90ce6.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-2f6618a016b45d26.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-3713889443fb7bea.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-677d6c5accc4e720.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-664f3e6bf23fe26d.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-fbedc56b0fc6fec3.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-80c24ebfb833516b.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-52b38c8b62b2f7a9.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-753fd57b99b33e72.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-ba50f8b8cf90eac5.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-f43c897f776e3ba4.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-1281b004ca04ec16.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-29711796eaedc6ea.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-f61490b3d499378b.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-48f913ee54104648.rlib" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-1a2195e5af3bb1e3.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/nix/store/dfrvks8s214ad1f3a1cdbnnbd0s490ni-rustc-1.77.2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/build/rustc-1.77.2-src/build/bootstrap/debug/build/proc-macro2-171ae2aa732da0bb/build_script_build-171ae2aa732da0bb" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs"
x86_64-unknown-linux-gnu-rustc>   = note: /nix/store/nd7ass1isid4vswpkfm82za4mcdwhzsm-binutils-2.41/bin/ld: cannot find -lgcc_s: No such file or directory
```

It looks like this also shows up trying to build static musl binaries, see: https://github.com/rust-lang/rust/issues/82521

And alpine appears to have a patch to work around this: https://github.com/rust-lang/rust/issues/55120

So maybe nixpkgs needs to patch the linking of libgcc_s out of the rust toolchain under `stdenv.targetPlatform.useLLVM`.

Reading the LSB, https://refspecs.linuxbase.org/LSB_4.1.0/LSB-Core-generic/LSB-Core-generic/libgcc-s.html the interfaces provided by `libgcc_s` are those relating to unwinding, which on `pkgsLLVM` is provided by `llvm-project/libunwind`.

Looks like enabling `llvm-libunwind` to either `in-tree` or `system` still results in this:
```
error: could not compile `typenum` (build script) due to 1 previous error
error: linking with `/nix/store/pjq4j90az01m6hgf7fd3jiilnp7f0cb6-aarch64-unknown-linux-gnu-clang-wrapper-17.0.6/bin/aarch64-unknown-linux-gnu-clang` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/bin:/build/rustc-1.78.0-src/build/aarch64-unknown-linux-gnu/stage0/bin:/nix/store/a11996pij79myrp43myn4zycdlzh9jsw-gcc-wrapper-13.3.0/bin:/nix/store/mcmyjzgqhfq6ycm0hz2482f454gd7ipc-gcc-13.3.0/bin:/nix/store/cv5w5wlyhfx1246dk4mdigj6x33w8jm2-glibc-2.39-52-bin/bin:/nix/store/84yg60swk80b04apprb1432kir41bvzj-coreutils-9.5/bin:/nix/store/cx2igw6n2pdfv62ksgk9vzxlk33rn6ak-binutils-wrapper-2.41/bin:/nix/store/y4fgprf79mjq30vc5wwz8pb15z2wpfl6-binutils-2.41/bin:/nix/store/c0x6ljv1blsrg8b26ga9afw2ijvb2kfw-pkg-config-wrapper-0.29.2/bin:/nix/store/ss6fxv2sc84ghnnmiik0i8p58z93n6pk-file-5.45/bin:/nix/store/q3x28mimkawkdjlvd78jxv3s0fk25vz8-python3-3.11.9/bin:/nix/store/l1zlwibr9j4avxy5fgljf8sifi2swxl0-rustc-wrapper-1.78.0/bin:/nix/store/dr2r1k4j95yphc49fcs2xsdqni2cxldn-cmake-3.29.3/bin:/nix/store/kgvi8kib1hknbc76539l9b1r7qaqrb30-which-2.21/bin:/nix/store/ir3q2pn7mgbazas1fgbhdhjaw35148ma-remove-references-to/bin:/nix/store/ihvflbs8s0dwlw0rjgxavz91fz4cgwhx-xz-5.4.6-bin/bin:/nix/store/2si2znpivnf74djn3m2shw6k3wrwn0ya-lndir-1.0.5/bin:/nix/store/dwh9z3c0rar7vc8n0s00hnbbljgy6flb-patchelf-0.15.0/bin:/nix/store/i0mly410wz8sn1xmvyh0dhk28hz0bvpd-openssl-3.0.13-bin/bin:/nix/store/xxlgyd2nj8dr32jhkriixzjy1pag1j84-llvm-18.1.7-dev/bin:/nix/store/bvlna0s53w3i62ndczah0ymvy3317kw5-ncurses-6.4.20221231-dev/bin:/nix/store/924k307m9myphvm8aqc82nmj44iqsqgl-ncurses-6.4.20221231/bin:/nix/store/apy4l0hvgm7q4zsrcizmg69dcf23c61k-llvm-18.1.7/bin:/nix/store/84yg60swk80b04apprb1432kir41bvzj-coreutils-9.5/bin:/nix/store/lxj7k5g73inw4h9jw7cw99jwig93fyqx-findutils-4.9.0/bin:/nix/store/fa6d8rg6lpk5g8g1rqcjrv9gm3gb8xn0-diffutils-3.10/bin:/nix/store/yxjfnwjn4azs2b5781ddnhmafmixpfic-gnused-4.9/bin:/nix/store/gy4xhd2wvf1m8v47k7iccjhg0x1d2898-gnugrep-3.11/bin:/nix/store/qnz36184dh0da9v0aagvasq709291wwi-gawk-5.2.2/bin:/nix/store/1cqyddnhggg6xcxni7k13pwlh3pygxyn-gnutar-1.35/bin:/nix/store/jd4i322cwm5b37b3rsgi4s08q34qjlll-gzip-1.13/bin:/nix/store/jmb4m219kv4v8n2hdmdjn23ih4f2vryf-bzip2-1.0.8-bin/bin:/nix/store/3w7ldfr5lrwmvfkhgmvgiq0dsb15f87a-gnumake-4.4.1/bin:/nix/store/gwqh47kdw6y1gv7mfwy0y00s692bs90f-bash-5.2p26/bin:/nix/store/cxkaib1mlvy4adahr01vp7j492x04n7h-patch-2.7.6/bin:/nix/store/ihvflbs8s0dwlw0rjgxavz91fz4cgwhx-xz-5.4.6-bin/bin:/nix/store/fdfc9dkg8njgvgpv8a39hawyicxc8kpa-file-5.45/bin" VSLANG="1033" "/nix/store/pjq4j90az01m6hgf7fd3jiilnp7f0cb6-aarch64-unknown-linux-gnu-clang-wrapper-17.0.6/bin/aarch64-unknown-linux-gnu-clang" "/build/rustcLY4Fmh/symbols.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.build_script_build.bf285b65bf450021-cgu.0.rcgu.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.build_script_build.bf285b65bf450021-cgu.1.rcgu.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.build_script_build.bf285b65bf450021-cgu.2.rcgu.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.build_script_build.bf285b65bf450021-cgu.3.rcgu.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.build_script_build.bf285b65bf450021-cgu.4.rcgu.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.build_script_build.bf285b65bf450021-cgu.5.rcgu.o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8.4wzy2y3g3v5n93k5.rcgu.o" "-Wl,--as-needed" "-L" "/build/rustc-1.78.0-src/build/bootstrap/debug/deps" "-L" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/build/rustc-1.78.0-src/build/bootstrap/debug/deps/libversion_check-0a854a65c02b3886.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd-9580f1bc6c74c617.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libpanic_unwind-788210a6c9c0ccad.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libobject-3abd70a15a19799d.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libmemchr-f89acf801e061d59.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libaddr2line-dac33b90a02a8d9f.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libgimli-e13380707f10595a.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/librustc_demangle-71da8908091bc0c2.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libstd_detect-169a89eab29e77c5.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libhashbrown-380b9767adbd299d.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-af90567433335a01.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libminiz_oxide-0d4d95d36505082d.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libadler-ea84e4d3e3e7101c.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libunwind-4b2ad54dfebf2bb6.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcfg_if-452a3ae7dcc6aec1.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/liblibc-5cf6311f20666b17.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/liballoc-9008138a274897a6.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/librustc_std_workspace_core-1f67b19bacc355b2.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcore-f41117d939b65698.rlib" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-15d71068531bf6d2.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/nix/store/c61javhyrbsc4pf5v941y0c1awhhahjy-rustc-1.78.0/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/build/rustc-1.78.0-src/build/bootstrap/debug/build/generic-array-9c3e49efb04b6eb8/build_script_build-9c3e49efb04b6eb8" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,--strip-debug" "-nodefaultlibs"
  = note: /nix/store/y4fgprf79mjq30vc5wwz8pb15z2wpfl6-binutils-2.41/bin/ld: cannot find -lgcc_s: No such file or director
```

--- a/pkgs/development/compilers/rust/1_79.nix
+++ b/pkgs/development/compilers/rust/1_79.nix
@@ -17,20 +17,45 @@
 , wrapRustcWith
 , llvmPackages_18, llvm_18
 } @ args:
-
+let
+  llvmSharedFor = pkgSet: pkgSet.llvmPackages_18.libllvm.override ({
+    enableSharedLibraries = true;
+  } // lib.optionalAttrs (stdenv.targetPlatform.useLLVM or false) {
+    # Force LLVM to compile using clang + LLVM libs when targeting pkgsLLVM
+    stdenv = pkgSet.stdenv.override { allowedRequisites = null; cc = pkgSet.llvmPackages_18.clangUseLLVM; };
+  });
+in
 import ./default.nix {
   rustcVersion = "1.79.0";
   rustcSha256 = "sha256-Fy7PPH0fnZ+xbNKmKIaXgmcEFt7QEp5SSoZ1H5YUSMA=";
 
-  llvmSharedForBuild = pkgsBuildBuild.llvmPackages_18.libllvm.override { enableSharedLibraries = true; };
-  llvmSharedForHost = pkgsBuildHost.llvmPackages_18.libllvm.override { enableSharedLibraries = true; };
-  llvmSharedForTarget = pkgsBuildTarget.llvmPackages_18.libllvm.override { enableSharedLibraries = true; };
+  llvmSharedForBuild = llvmSharedFor pkgsBuildBuild;
+  llvmSharedForHost = llvmSharedFor pkgsBuildHost;
+  llvmSharedForTarget = llvmSharedFor pkgsBuildTarget;
 
   # For use at runtime
-  llvmShared = llvm_18.override { enableSharedLibraries = true; };
+  llvmShared = llvmSharedFor { inherit llvmPackages_18 stdenv; };
 
   # Expose llvmPackages used for rustc from rustc via passthru for LTO in Firefox
-  llvmPackages = llvmPackages_18;
+  llvmPackages = if (stdenv.targetPlatform.useLLVM or false) then (let
+    setStdenv = pkg: pkg.override {
+      stdenv = stdenv.override { allowedRequisites = null; cc = llvmPackages_18.clangUseLLVM; };
+    };
+  in rec {
+    libunwind = setStdenv llvmPackages_18.libunwind;
+    llvm = setStdenv llvmPackages_18.llvm;
+
+    libcxx = llvmPackages_18.libcxx.override {
+      stdenv = stdenv.override {
+        allowedRequisites = null;
+        cc = llvmPackages_18.clangNoLibcxx;
+        hostPlatform = stdenv.hostPlatform // {
+          useLLVM = !stdenv.isDarwin;
+        };
+      };
+      inherit libunwind;
+    };
+  }) else llvmPackages_18;
 
   # Note: the version MUST be one version prior to the version we're
   # building

--- a/pkgs/development/compilers/rust/rustc.nix
+++ b/pkgs/development/compilers/rust/rustc.nix
@@ -1,6 +1,6 @@
 { lib, stdenv, removeReferencesTo, pkgsBuildBuild, pkgsBuildHost, pkgsBuildTarget, targetPackages
 , llvmShared, llvmSharedForBuild, llvmSharedForHost, llvmSharedForTarget, llvmPackages
-, fetchurl, file, python3
+, runCommandLocal, fetchurl, file, python3
 , darwin, cargo, cmake, rustc, rustfmt
 , pkg-config, openssl, xz
 , libiconv
@@ -24,6 +24,7 @@
 let
   inherit (lib) optionals optional optionalString concatStringsSep;
   inherit (darwin.apple_sdk.frameworks) Security;
+  useLLVM = stdenv.targetPlatform.useLLVM or false;
 in stdenv.mkDerivation (finalAttrs: {
   pname = "${targetPackages.stdenv.cc.targetPrefix}rustc";
   inherit version;
@@ -66,14 +67,16 @@ in stdenv.mkDerivation (finalAttrs: {
        # when linking stage1 libstd: cc: undefined reference to `__cxa_begin_catch'
        # This doesn't apply to cross-building for FreeBSD because the host
        # uses libstdc++, but the target (used for building std) uses libc++
-       optional (stdenv.isLinux && !withBundledLLVM && !stdenv.targetPlatform.isFreeBSD) "--push-state --as-needed -lstdc++ --pop-state"
+      optional (stdenv.isLinux && !withBundledLLVM && !stdenv.targetPlatform.isFreeBSD && !useLLVM)
+        "--push-state --as-needed -lstdc++ --pop-state"
+    ++ optional (stdenv.isLinux && !withBundledLLVM && !stdenv.targetPlatform.isFreeBSD && useLLVM)
+        "--push-state --as-needed -L${llvmPackages.libcxx}/lib -lc++ -lc++abi -lLLVM-${lib.versions.major llvmPackages.llvm.version} --pop-state"
     ++ optional (stdenv.isDarwin && !withBundledLLVM) "-lc++ -lc++abi"
     ++ optional stdenv.isFreeBSD "-rpath ${llvmPackages.libunwind}/lib"
-    ++ optional stdenv.isDarwin "-rpath ${llvmSharedForHost}/lib");
+    ++ optional stdenv.isDarwin "-rpath ${llvmSharedForHost.lib}/lib");
 
   # Increase codegen units to introduce parallelism within the compiler.
   RUSTFLAGS = "-Ccodegen-units=10";
-
   RUSTDOCFLAGS = "-A rustdoc::broken-intra-doc-links";
 
   # We need rust to build rust. If we don't provide it, configure will try to download it.
@@ -152,7 +155,7 @@ in stdenv.mkDerivation (finalAttrs: {
     # Since fastCross only builds std, it doesn't make sense (and
     # doesn't work) to build a linker.
     "--disable-llvm-bitcode-linker"
-  ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox) [
+  ] ++ optionals (stdenv.isLinux && !stdenv.targetPlatform.isRedox && !(stdenv.targetPlatform.useLLVM or false)) [
     "--enable-profiler" # build libprofiler_builtins
   ] ++ optionals stdenv.buildPlatform.isMusl [
     "${setBuild}.musl-root=${pkgsBuildBuild.targetPackages.stdenv.cc.libc}"
@@ -165,6 +168,10 @@ in stdenv.mkDerivation (finalAttrs: {
   ] ++ optionals (stdenv.isDarwin && stdenv.isx86_64) [
     # https://github.com/rust-lang/rust/issues/92173
     "--set rust.jemalloc"
+  ] ++ optionals useLLVM [
+    # https://github.com/NixOS/nixpkgs/issues/311930
+    "--llvm-libunwind=${if withBundledLLVM then "in-tree" else "system"}"
+    "--enable-use-libcxx"
   ];
 
   # if we already have a rust compiler for build just compile the target std
@@ -189,6 +196,7 @@ in stdenv.mkDerivation (finalAttrs: {
     python ./x.py --keep-stage=0 --stage=1 install library/std
     mkdir -v $out/bin $doc $man
     ln -s ${rustc.unwrapped}/bin/{rustc,rustdoc} $out/bin
+    rm -rf -v $out/lib/rustlib/{manifest-rust-std-,}${stdenv.hostPlatform.rust.rustcTargetSpec}
     ln -s ${rustc.unwrapped}/lib/rustlib/{manifest-rust-std-,}${stdenv.hostPlatform.rust.rustcTargetSpec} $out/lib/rustlib/
     echo rust-std-${stdenv.hostPlatform.rust.rustcTargetSpec} >> $out/lib/rustlib/components
     lndir ${rustc.doc} $doc
@@ -248,7 +256,16 @@ in stdenv.mkDerivation (finalAttrs: {
 
   buildInputs = [ openssl ]
     ++ optionals stdenv.isDarwin [ libiconv Security ]
-    ++ optional (!withBundledLLVM) llvmShared;
+    ++ optional (!withBundledLLVM) llvmShared.lib
+    ++ optional (useLLVM && !withBundledLLVM) [
+      llvmPackages.libunwind
+      # Hack which is used upstream https://github.com/gentoo/gentoo/blob/master/dev-lang/rust/rust-1.78.0.ebuild#L284
+      (runCommandLocal "libunwind-libgcc" {} ''
+        mkdir -p $out/lib
+        ln -s ${llvmPackages.libunwind}/lib/libunwind.so $out/lib/libgcc_s.so
+        ln -s ${llvmPackages.libunwind}/lib/libunwind.so $out/lib/libgcc_s.so.1
+      '')
+    ];
 
   outputs = [ "out" "man" "doc" ];
   setOutputFlags = false;

